File: macos/ScreenControl/TestServer.m
//
//  TestServer.m
//  ScreenControl
//
//  Remote testing server for automated agent testing.
//  Only available in DEBUG builds.
//

#import "TestServer.h"

#ifdef DEBUG

#import "AppDelegate.h"
#import <sys/socket.h>
#import <netinet/in.h>
#import <arpa/inet.h>

@interface TestServer ()
@property (nonatomic, assign) int serverSocket;
@property (nonatomic, strong) dispatch_source_t acceptSource;
@property (nonatomic, assign) BOOL isRunning;
@property (nonatomic, assign) uint16_t port;
@end

@implementation TestServer

- (instancetype)initWithAppDelegate:(AppDelegate *)appDelegate {
    self = [super init];
    if (self) {
        _appDelegate = appDelegate;
        _serverSocket = -1;
        _isRunning = NO;
        _port = 0;
    }
    return self;
}

- (void)dealloc {
    [self stop];
}

- (BOOL)startOnPort:(uint16_t)port {
    if (self.isRunning) {
        NSLog(@"[TestServer] Already running on port %d", self.port);
        return YES;
    }

    // Try primary port, then fallback
    uint16_t ports[] = { port, (uint16_t)(port + 1) };

    for (int i = 0; i < 2; i++) {
        if ([self bindToPort:ports[i]]) {
            self.port = ports[i];
            self.isRunning = YES;
            NSLog(@"[TestServer] Started on 0.0.0.0:%d (DEBUG BUILD ONLY - accessible remotely)", self.port);
            return YES;
        }
    }

    NSLog(@"[TestServer] Failed to bind to any port");
    return NO;
}

- (BOOL)bindToPort:(uint16_t)port {
    // Create socket
    self.serverSocket = socket(AF_INET, SOCK_STREAM, 0);
    if (self.serverSocket < 0) {
        NSLog(@"[TestServer] Failed to create socket: %s", strerror(errno));
        return NO;
    }

    // Allow address reuse
    int optval = 1;
    setsockopt(self.serverSocket, SOL_SOCKET, SO_REUSEADDR, &optval, sizeof(optval));

    // Bind to all interfaces for remote testing (DEBUG builds only)
    struct sockaddr_in addr;
    memset(&addr, 0, sizeof(addr));
    addr.sin_family = AF_INET;
    addr.sin_port = htons(port);
    addr.sin_addr.s_addr = INADDR_ANY;  // 0.0.0.0 - allows remote access in DEBUG builds

    if (bind(self.serverSocket, (struct sockaddr *)&addr, sizeof(addr)) < 0) {
        NSLog(@"[TestServer] Failed to bind to port %d: %s", port, strerror(errno));
        close(self.serverSocket);
        self.serverSocket = -1;
        return NO;
    }

    // Listen
    if (listen(self.serverSocket, 5) < 0) {
        NSLog(@"[TestServer] Failed to listen: %s", strerror(errno));
        close(self.serverSocket);
        self.serverSocket = -1;
        return NO;
    }

    // Create dispatch source for accepting connections
    self.acceptSource = dispatch_source_create(DISPATCH_SOURCE_TYPE_READ, self.serverSocket, 0, dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0));

    __weak typeof(self) weakSelf = self;
    dispatch_source_set_event_handler(self.acceptSource, ^{
        [weakSelf acceptConnection];
    });

    dispatch_source_set_cancel_handler(self.acceptSource, ^{
        if (weakSelf.serverSocket >= 0) {
            close(weakSelf.serverSocket);
            weakSelf.serverSocket = -1;
        }
    });

    dispatch_resume(self.acceptSource);

    return YES;
}

- (void)stop {
    if (!self.isRunning) return;

    NSLog(@"[TestServer] Stopping...");

    if (self.acceptSource) {
        dispatch_source_cancel(self.acceptSource);
        self.acceptSource = nil;
    }

    self.isRunning = NO;
    self.port = 0;
}

- (void)acceptConnection {
    struct sockaddr_in clientAddr;
    socklen_t clientLen = sizeof(clientAddr);

    int clientSocket = accept(self.serverSocket, (struct sockaddr *)&clientAddr, &clientLen);
    if (clientSocket < 0) {
        NSLog(@"[TestServer] Accept failed: %s", strerror(errno));
        return;
    }

    // Handle connection in background
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        [self handleClient:clientSocket];
    });
}

- (void)handleClient:(int)clientSocket {
    // Read HTTP request
    char buffer[4096];
    ssize_t bytesRead = recv(clientSocket, buffer, sizeof(buffer) - 1, 0);

    if (bytesRead <= 0) {
        close(clientSocket);
        return;
    }

    buffer[bytesRead] = '\0';
    NSString *request = [NSString stringWithUTF8String:buffer];

    // Parse HTTP request
    NSString *response = [self handleRequest:request];

    // Send HTTP response
    NSString *httpResponse = [NSString stringWithFormat:
        @"HTTP/1.1 200 OK\r\n"
        @"Content-Type: application/json\r\n"
        @"Access-Control-Allow-Origin: *\r\n"
        @"Connection: close\r\n"
        @"Content-Length: %lu\r\n"
        @"\r\n"
        @"%@",
        (unsigned long)[response lengthOfBytesUsingEncoding:NSUTF8StringEncoding],
        response
    ];

    const char *responseBytes = [httpResponse UTF8String];
    send(clientSocket, responseBytes, strlen(responseBytes), 0);

    close(clientSocket);
}

- (NSString *)handleRequest:(NSString *)httpRequest {
    // Handle GET requests first (before checking for body)
    if ([httpRequest hasPrefix:@"GET "]) {
        if ([httpRequest hasPrefix:@"GET /ping"]) {
            return [self handleMethod:@"ping" params:nil];
        } else if ([httpRequest hasPrefix:@"GET / "] || [httpRequest hasPrefix:@"GET / HTTP"] || [httpRequest hasPrefix:@"GET /\r\n"]) {
            // Base endpoint - return menu
            return [self handleMenu];
        }
    }

    // Handle OPTIONS (CORS preflight)
    if ([httpRequest hasPrefix:@"OPTIONS"]) {
        return @"{}";
    }

    // Extract JSON body from HTTP request
    NSRange bodyRange = [httpRequest rangeOfString:@"\r\n\r\n"];
    if (bodyRange.location == NSNotFound) {
        return @"{\"error\":\"Invalid request\"}";
    }

    NSString *body = [httpRequest substringFromIndex:NSMaxRange(bodyRange)];

    // Parse JSON
    NSData *jsonData = [body dataUsingEncoding:NSUTF8StringEncoding];
    if (!jsonData || jsonData.length == 0) {
        return @"{\"error\":\"Empty body\"}";
    }

    NSError *error;
    NSDictionary *json = [NSJSONSerialization JSONObjectWithData:jsonData options:0 error:&error];
    if (error || !json) {
        return [NSString stringWithFormat:@"{\"error\":\"Invalid JSON: %@\"}", error.localizedDescription];
    }

    NSString *method = json[@"method"];
    NSDictionary *params = json[@"params"];

    if (!method) {
        return @"{\"error\":\"Missing method\"}";
    }

    return [self handleMethod:method params:params];
}

- (NSString *)handleMethod:(NSString *)method params:(NSDictionary *)params {
    NSDictionary *result;

    if ([method isEqualToString:@"ping"]) {
        result = [self handlePing];
    } else if ([method isEqualToString:@"getState"]) {
        result = [self handleGetState];
    } else if ([method isEqualToString:@"getFields"]) {
        result = [self handleGetFields];
    } else if ([method isEqualToString:@"setField"]) {
        result = [self handleSetField:params];
    } else if ([method isEqualToString:@"clickButton"]) {
        result = [self handleClickButton:params];
    } else if ([method isEqualToString:@"getLogs"]) {
        result = [self handleGetLogs:params];
    } else if ([method isEqualToString:@"quit"]) {
        result = [self handleQuit];
    } else if ([method isEqualToString:@"restart"]) {
        result = [self handleRestart];
    } else if ([method isEqualToString:@"getVersion"]) {
        result = [self handleGetVersion];
    } else {
        result = @{@"error": [NSString stringWithFormat:@"Unknown method: %@", method]};
    }

    NSData *jsonData = [NSJSONSerialization dataWithJSONObject:result options:0 error:nil];
    return [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];
}

#pragma mark - Method Handlers

- (NSString *)handleMenu {
    NSDictionary *menu = @{
        @"service": @"ScreenControl Test Server",
        @"description": @"HTTP API for automated testing of ScreenControl agent (DEBUG builds only)",
        @"version": [[NSBundle mainBundle] objectForInfoDictionaryKey:@"CFBundleShortVersionString"] ?: @"1.0.0",
        @"port": @(self.port),
        @"usage": @"Send POST requests with JSON body: {\"method\": \"<method_name>\", \"params\": {...}}",
        @"availableCommands": @[
            @{
                @"method": @"ping",
                @"description": @"Health check - returns server status",
                @"params": @{},
                @"example": @{@"method": @"ping"}
            },
            @{
                @"method": @"getState",
                @"description": @"Get current connection state and status",
                @"params": @{},
                @"example": @{@"method": @"getState"}
            },
            @{
                @"method": @"getFields",
                @"description": @"Get all UI field values",
                @"params": @{},
                @"example": @{@"method": @"getFields"}
            },
            @{
                @"method": @"setField",
                @"description": @"Set a UI field value",
                @"params": @{@"field": @"<field_name>", @"value": @"<value>"},
                @"example": @{@"method": @"setField", @"params": @{@"field": @"serverUrl", @"value": @"ws://localhost:3000"}}
            },
            @{
                @"method": @"clickButton",
                @"description": @"Click a UI button",
                @"params": @{@"button": @"<button_name>"},
                @"availableButtons": @[
                    @"connect", @"disconnect", @"saveSettings", @"reconnect",
                    @"discoverAndJoin", @"startBrowserBridge", @"stopBrowserBridge"
                ],
                @"example": @{@"method": @"clickButton", @"params": @{@"button": @"startBrowserBridge"}}
            },
            @{
                @"method": @"getLogs",
                @"description": @"Get recent debug logs",
                @"params": @{@"lines": @"<number_of_lines>"},
                @"example": @{@"method": @"getLogs", @"params": @{@"lines": @50}}
            },
            @{
                @"method": @"getVersion",
                @"description": @"Get application version information",
                @"params": @{},
                @"example": @{@"method": @"getVersion"}
            },
            @{
                @"method": @"restart",
                @"description": @"Restart the application",
                @"params": @{},
                @"example": @{@"method": @"restart"}
            },
            @{
                @"method": @"quit",
                @"description": @"Quit the application",
                @"params": @{},
                @"example": @{@"method": @"quit"}
            }
        ]
    };

    NSData *jsonData = [NSJSONSerialization dataWithJSONObject:menu options:NSJSONWritingPrettyPrinted error:nil];
    return [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];
}

- (NSDictionary *)handlePing {
    NSString *version = [[NSBundle mainBundle] objectForInfoDictionaryKey:@"CFBundleShortVersionString"] ?: @"1.0.0";
    return @{
        @"pong": @YES,
        @"version": version,
        @"debug": @YES,
        @"port": @(self.port)
    };
}

- (NSDictionary *)handleGetState {
    __block NSDictionary *state;

    dispatch_sync(dispatch_get_main_queue(), ^{
        AppDelegate *app = self.appDelegate;
        state = @{
            @"connected": @(app.debugIsConnected),
            @"serverUrl": app.debugServerUrlField.stringValue ?: @"",
            @"endpointUuid": app.debugEndpointUuidField.stringValue ?: @"",
            @"customerId": app.debugCustomerIdField.stringValue ?: @"",
            @"connectionStatus": app.debugConnectionStatusLabel.stringValue ?: @"Unknown",
            // NOTE: Auto-reconnect is now handled by the service
            @"autoReconnectEnabled": @(YES),
            @"reconnectAttempt": @(0),
            @"reconnectPending": @(NO)
        };
    });

    return state;
}

- (NSDictionary *)handleGetFields {
    __block NSDictionary *fields;

    dispatch_sync(dispatch_get_main_queue(), ^{
        AppDelegate *app = self.appDelegate;
        fields = @{
            @"serverUrl": app.debugServerUrlField.stringValue ?: @"",
            @"endpointUuid": app.debugEndpointUuidField.stringValue ?: @"",
            @"customerId": app.debugCustomerIdField.stringValue ?: @"",
            @"connectOnStartup": @(app.debugConnectOnStartupCheckbox.state == NSControlStateValueOn),
            @"mcpUrl": app.debugMcpUrlField.stringValue ?: @"",
            @"oauthStatus": app.debugOAuthStatusLabel.stringValue ?: @"",
            // Control Server (General tab) fields
            @"controlServerUrl": app.controlServerAddressField.stringValue ?: @"",
            @"controlServerStatus": app.connectionStatusLabel.stringValue ?: @""
        };
    });

    return fields;
}

- (NSDictionary *)handleSetField:(NSDictionary *)params {
    NSString *field = params[@"field"];
    NSString *value = params[@"value"];

    if (!field || !value) {
        return @{@"error": @"Missing field or value"};
    }

    __block BOOL success = NO;

    dispatch_sync(dispatch_get_main_queue(), ^{
        AppDelegate *app = self.appDelegate;

        if ([field isEqualToString:@"serverUrl"]) {
            app.debugServerUrlField.stringValue = value;
            success = YES;
        } else if ([field isEqualToString:@"endpointUuid"]) {
            app.debugEndpointUuidField.stringValue = value;
            success = YES;
        } else if ([field isEqualToString:@"customerId"]) {
            app.debugCustomerIdField.stringValue = value;
            success = YES;
        } else if ([field isEqualToString:@"connectOnStartup"]) {
            app.debugConnectOnStartupCheckbox.state = [value boolValue] ? NSControlStateValueOn : NSControlStateValueOff;
            success = YES;
        } else if ([field isEqualToString:@"mcpUrl"]) {
            app.debugMcpUrlField.stringValue = value;
            success = YES;
        } else if ([field isEqualToString:@"controlServerUrl"]) {
            app.controlServerAddressField.stringValue = value;
            success = YES;
        }
    });

    if (success) {
        return @{@"success": @YES, @"field": field, @"value": value};
    } else {
        return @{@"error": [NSString stringWithFormat:@"Unknown field: %@", field]};
    }
}

- (NSDictionary *)handleClickButton:(NSDictionary *)params {
    NSString *button = params[@"button"];

    if (!button) {
        return @{@"error": @"Missing button parameter"};
    }

    __block BOOL success = NO;
    __block NSString *action = @"";

    dispatch_sync(dispatch_get_main_queue(), ^{
        AppDelegate *app = self.appDelegate;

        if ([button isEqualToString:@"connect"]) {
            if (app.debugConnectButton.enabled) {
                [app debugConnectClicked:nil];
                action = @"connect";
                success = YES;
            }
        } else if ([button isEqualToString:@"disconnect"]) {
            if (app.debugDisconnectButton.enabled) {
                [app debugDisconnectClicked:nil];
                action = @"disconnect";
                success = YES;
            }
        } else if ([button isEqualToString:@"reconnect"]) {
            // Force reconnect - always available
            [app debugReconnectClicked:nil];
            action = @"reconnect";
            success = YES;
        } else if ([button isEqualToString:@"saveSettings"]) {
            [app debugSaveSettingsClicked:nil];
            action = @"saveSettings";
            success = YES;
        } else if ([button isEqualToString:@"discoverAndJoin"]) {
            // Trigger OAuth discovery and join
            [app discoverAndJoinClicked:nil];
            action = @"discoverAndJoin";
            success = YES;
        } else if ([button isEqualToString:@"controlServerConnect"]) {
            // Connect to control server (General tab)
            if (app.connectButton.enabled) {
                [app connectControlServer:nil];
                action = @"controlServerConnect";
                success = YES;
            }
        } else if ([button isEqualToString:@"startBrowserBridge"]) {
            // Start browser bridge server
            [app startBrowserBridge];
            action = @"startBrowserBridge";
            success = YES;
        } else if ([button isEqualToString:@"stopBrowserBridge"]) {
            // Stop browser bridge server
            [app stopBrowserBridge];
            action = @"stopBrowserBridge";
            success = YES;
        }
    });

    if (success) {
        return @{@"success": @YES, @"action": action};
    } else {
        return @{@"error": [NSString stringWithFormat:@"Unknown or disabled button: %@", button]};
    }
}

- (NSDictionary *)handleGetLogs:(NSDictionary *)params {
    NSInteger limit = [params[@"limit"] integerValue] ?: 50;

    __block NSString *logs = @"";

    dispatch_sync(dispatch_get_main_queue(), ^{
        AppDelegate *app = self.appDelegate;
        logs = app.debugLogTextView.string ?: @"";
    });

    // Get last N lines
    NSArray *lines = [logs componentsSeparatedByString:@"\n"];
    NSInteger start = MAX(0, (NSInteger)lines.count - limit);
    NSArray *recentLines = [lines subarrayWithRange:NSMakeRange(start, lines.count - start)];

    return @{
        @"logs": recentLines,
        @"total": @(lines.count),
        @"returned": @(recentLines.count)
    };
}

- (NSDictionary *)handleQuit {
    NSLog(@"[TestServer] Quit requested via test server");

    dispatch_async(dispatch_get_main_queue(), ^{
        [[NSApplication sharedApplication] terminate:nil];
    });

    return @{@"success": @YES, @"action": @"quit"};
}

- (NSDictionary *)handleRestart {
    NSLog(@"[TestServer] Restart requested via test server");

    // Get path to this app
    NSString *appPath = [[NSBundle mainBundle] bundlePath];

    dispatch_async(dispatch_get_main_queue(), ^{
        // Launch new instance after a short delay
        dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.5 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
            [[NSWorkspace sharedWorkspace] openApplicationAtURL:[NSURL fileURLWithPath:appPath]
                                                  configuration:[NSWorkspaceOpenConfiguration configuration]
                                              completionHandler:nil];
        });

        // Terminate current instance
        [[NSApplication sharedApplication] terminate:nil];
    });

    return @{@"success": @YES, @"action": @"restart"};
}

- (NSDictionary *)handleGetVersion {
    NSBundle *bundle = [NSBundle mainBundle];

    // Get version info from bundle
    NSString *version = [bundle objectForInfoDictionaryKey:@"CFBundleShortVersionString"] ?: @"1.0.0";
    NSString *buildNumber = [bundle objectForInfoDictionaryKey:@"CFBundleVersion"] ?: @"1";

    // Get build date from executable modification date
    NSString *executablePath = [bundle executablePath];
    NSDictionary *attrs = [[NSFileManager defaultManager] attributesOfItemAtPath:executablePath error:nil];
    NSDate *buildDate = attrs[NSFileModificationDate] ?: [NSDate date];
    NSDateFormatter *formatter = [[NSDateFormatter alloc] init];
    formatter.dateFormat = @"yyyy-MM-dd";
    NSString *buildDateString = [formatter stringFromDate:buildDate];

    // Get architecture
    NSString *arch = @"unknown";
#if defined(__arm64__)
    arch = @"arm64";
#elif defined(__x86_64__)
    arch = @"x86_64";
#endif

    // Git commit would need to be embedded at build time via a build script
    // For now, check if we have it in Info.plist
    NSString *gitCommit = [bundle objectForInfoDictionaryKey:@"GitCommit"] ?: @"unknown";

    // Get app startup time (uptime)
    static NSDate *startTime = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        startTime = [NSDate date];
    });
    NSTimeInterval uptime = [[NSDate date] timeIntervalSinceDate:startTime];

    return @{
        @"version": version,
        @"build": buildNumber,
        @"buildDate": buildDateString,
        @"gitCommit": gitCommit,
        @"platform": @"macos",
        @"arch": arch,
        @"uptime": @((NSInteger)uptime)
    };
}

@end

#endif // DEBUG

File: linux/screencontrol/test_server.cpp
/**
 * TestServer Implementation
 *
 * Simple HTTP server using POSIX sockets for automated testing.
 */

#include "test_server.h"
#include "core/logger.h"
#include "libs/json.hpp"

#include <iostream>
#include <sstream>
#include <cstring>
#include <ctime>
#include <chrono>

#include <sys/socket.h>
#include <sys/types.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <unistd.h>
#include <fcntl.h>
#include <sys/stat.h>
#include <sys/utsname.h>

using json = nlohmann::json;

namespace ScreenControl
{

static std::chrono::steady_clock::time_point g_startTime;

TestServer::TestServer()
{
    g_startTime = std::chrono::steady_clock::now();
}

TestServer::~TestServer()
{
    stop();
}

bool TestServer::start(uint16_t port)
{
    if (m_running) return true;

    // Try primary port, then fallback
    uint16_t ports[] = { port, static_cast<uint16_t>(port + 1) };

    for (auto p : ports)
    {
        m_serverSocket = socket(AF_INET, SOCK_STREAM, 0);
        if (m_serverSocket < 0)
        {
            Logger::error("TestServer: Failed to create socket");
            continue;
        }

        // Allow address reuse
        int optval = 1;
        setsockopt(m_serverSocket, SOL_SOCKET, SO_REUSEADDR, &optval, sizeof(optval));

        // Bind to localhost ONLY
        struct sockaddr_in addr;
        memset(&addr, 0, sizeof(addr));
        addr.sin_family = AF_INET;
        addr.sin_port = htons(p);
        addr.sin_addr.s_addr = inet_addr("127.0.0.1");

        if (bind(m_serverSocket, reinterpret_cast<struct sockaddr*>(&addr), sizeof(addr)) < 0)
        {
            Logger::error("TestServer: Failed to bind to port " + std::to_string(p));
            close(m_serverSocket);
            m_serverSocket = -1;
            continue;
        }

        if (listen(m_serverSocket, 5) < 0)
        {
            Logger::error("TestServer: Failed to listen");
            close(m_serverSocket);
            m_serverSocket = -1;
            continue;
        }

        m_port = p;
        m_running = true;

        Logger::info("TestServer: Started on localhost:" + std::to_string(m_port) + " (DEBUG BUILD ONLY)");

        m_listenThread = std::thread(&TestServer::listenLoop, this);
        return true;
    }

    Logger::error("TestServer: Failed to bind to any port");
    return false;
}

void TestServer::stop()
{
    if (!m_running) return;

    Logger::info("TestServer: Stopping...");

    m_running = false;

    if (m_serverSocket >= 0)
    {
        shutdown(m_serverSocket, SHUT_RDWR);
        close(m_serverSocket);
        m_serverSocket = -1;
    }

    if (m_listenThread.joinable())
    {
        m_listenThread.join();
    }

    m_port = 0;
}

void TestServer::listenLoop()
{
    while (m_running)
    {
        struct sockaddr_in clientAddr;
        socklen_t clientLen = sizeof(clientAddr);

        int clientSocket = accept(m_serverSocket, reinterpret_cast<struct sockaddr*>(&clientAddr), &clientLen);

        if (clientSocket < 0)
        {
            if (m_running)
            {
                Logger::error("TestServer: Accept failed");
            }
            continue;
        }

        // Handle in same thread (simple approach)
        handleClient(clientSocket);
    }
}

void TestServer::handleClient(int clientSocket)
{
    char buffer[4096];
    ssize_t bytesRead = recv(clientSocket, buffer, sizeof(buffer) - 1, 0);

    if (bytesRead <= 0)
    {
        close(clientSocket);
        return;
    }

    buffer[bytesRead] = '\0';
    std::string request(buffer);

    std::string responseBody = handleRequest(request);

    std::ostringstream httpResponse;
    httpResponse << "HTTP/1.1 200 OK\r\n"
                 << "Content-Type: application/json\r\n"
                 << "Access-Control-Allow-Origin: *\r\n"
                 << "Connection: close\r\n"
                 << "Content-Length: " << responseBody.size() << "\r\n"
                 << "\r\n"
                 << responseBody;

    std::string response = httpResponse.str();
    send(clientSocket, response.c_str(), response.size(), 0);

    close(clientSocket);
}

std::string TestServer::handleRequest(const std::string& httpRequest)
{
    // Find body
    size_t bodyPos = httpRequest.find("\r\n\r\n");
    if (bodyPos == std::string::npos)
    {
        // Check for GET /ping
        if (httpRequest.find("GET /ping") == 0)
        {
            return handleMethod("ping", "{}");
        }
        return R"({"error":"Invalid request"})";
    }

    // Handle OPTIONS (CORS preflight)
    if (httpRequest.find("OPTIONS") == 0)
    {
        return "{}";
    }

    std::string body = httpRequest.substr(bodyPos + 4);
    if (body.empty())
    {
        return R"({"error":"Empty body"})";
    }

    try
    {
        json j = json::parse(body);

        if (!j.contains("method"))
        {
            return R"({"error":"Missing method"})";
        }

        std::string method = j["method"];
        std::string params = j.contains("params") ? j["params"].dump() : "{}";

        return handleMethod(method, params);
    }
    catch (const std::exception& e)
    {
        return R"({"error":"Invalid JSON: )" + std::string(e.what()) + R"("})";
    }
}

std::string TestServer::handleMethod(const std::string& method, const std::string& params)
{
    if (method == "ping") return handlePing();
    if (method == "getState") return handleGetState();
    if (method == "getFields") return handleGetFields();
    if (method == "setField") return handleSetField(params);
    if (method == "clickButton") return handleClickButton(params);
    if (method == "getLogs") return handleGetLogs(params);
    if (method == "getVersion") return handleGetVersion();
    if (method == "quit") return handleQuit();
    if (method == "restart") return handleRestart();

    return R"({"error":"Unknown method: )" + method + R"("})";
}

std::string TestServer::handlePing()
{
    json result;
    result["pong"] = true;
    result["version"] = VERSION;
    result["debug"] = true;
    result["port"] = m_port;
    return result.dump();
}

std::string TestServer::handleGetState()
{
    json result;

    if (m_getStatusCallback)
    {
        auto [connected, status] = m_getStatusCallback();
        result["connected"] = connected;
        result["connectionStatus"] = status;
    }
    else
    {
        result["connected"] = false;
        result["connectionStatus"] = "Unknown";
    }

    if (m_getFieldCallback)
    {
        result["serverUrl"] = m_getFieldCallback("serverUrl");
        result["endpointUuid"] = m_getFieldCallback("endpointUuid");
        result["customerId"] = m_getFieldCallback("customerId");
    }

    return result.dump();
}

std::string TestServer::handleGetFields()
{
    json result;

    if (m_getFieldCallback)
    {
        result["serverUrl"] = m_getFieldCallback("serverUrl");
        result["endpointUuid"] = m_getFieldCallback("endpointUuid");
        result["customerId"] = m_getFieldCallback("customerId");
        result["connectOnStartup"] = m_getFieldCallback("connectOnStartup") == "true";
    }

    return result.dump();
}

std::string TestServer::handleSetField(const std::string& params)
{
    try
    {
        json j = json::parse(params);

        std::string field = j.value("field", "");
        std::string value = j.value("value", "");

        if (field.empty())
        {
            return R"({"error":"Missing field"})";
        }

        if (m_setFieldCallback && m_setFieldCallback(field, value))
        {
            json result;
            result["success"] = true;
            result["field"] = field;
            result["value"] = value;
            return result.dump();
        }

        return R"({"error":"Unknown field: )" + field + R"("})";
    }
    catch (...)
    {
        return R"({"error":"Invalid params"})";
    }
}

std::string TestServer::handleClickButton(const std::string& params)
{
    try
    {
        json j = json::parse(params);
        std::string button = j.value("button", "");

        if (button.empty())
        {
            return R"({"error":"Missing button parameter"})";
        }

        if (m_clickButtonCallback && m_clickButtonCallback(button))
        {
            json result;
            result["success"] = true;
            result["action"] = button;
            return result.dump();
        }

        return R"({"error":"Unknown or disabled button: )" + button + R"("})";
    }
    catch (...)
    {
        return R"({"error":"Invalid params"})";
    }
}

std::string TestServer::handleGetLogs(const std::string& params)
{
    int limit = 50;
    try
    {
        json j = json::parse(params);
        limit = j.value("limit", 50);
    }
    catch (...) {}

    json result;
    result["logs"] = json::array();

    if (m_getLogsCallback)
    {
        auto logs = m_getLogsCallback(limit);
        result["logs"] = logs;
        result["total"] = logs.size();
        result["returned"] = logs.size();
    }
    else
    {
        result["total"] = 0;
        result["returned"] = 0;
    }

    return result.dump();
}

std::string TestServer::handleGetVersion()
{
    json result;
    result["version"] = VERSION;
    result["build"] = "1";

    // Get build date from executable
    struct stat attr;
    if (stat("/proc/self/exe", &attr) == 0)
    {
        char dateBuf[32];
        strftime(dateBuf, sizeof(dateBuf), "%Y-%m-%d", localtime(&attr.st_mtime));
        result["buildDate"] = dateBuf;
    }
    else
    {
        result["buildDate"] = "unknown";
    }

    result["gitCommit"] = "unknown";
    result["platform"] = "linux";

#if defined(__x86_64__)
    result["arch"] = "x64";
#elif defined(__aarch64__)
    result["arch"] = "arm64";
#else
    result["arch"] = "unknown";
#endif

    auto now = std::chrono::steady_clock::now();
    auto uptime = std::chrono::duration_cast<std::chrono::seconds>(now - g_startTime).count();
    result["uptime"] = uptime;

    return result.dump();
}

std::string TestServer::handleQuit()
{
    Logger::info("TestServer: Quit requested via test server");

    if (m_quitCallback)
    {
        m_quitCallback();
    }

    json result;
    result["success"] = true;
    result["action"] = "quit";
    return result.dump();
}

std::string TestServer::handleRestart()
{
    Logger::info("TestServer: Restart requested via test server");

    // Get path to current executable
    char exePath[1024];
    ssize_t len = readlink("/proc/self/exe", exePath, sizeof(exePath) - 1);
    if (len > 0)
    {
        exePath[len] = '\0';

        // Fork and exec
        pid_t pid = fork();
        if (pid == 0)
        {
            // Child - wait a bit then exec
            usleep(500000);
            execl(exePath, exePath, nullptr);
            exit(1);
        }
    }

    // Quit this instance
    if (m_quitCallback)
    {
        m_quitCallback();
    }

    json result;
    result["success"] = true;
    result["action"] = "restart";
    return result.dump();
}

} // namespace ScreenControl
